src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,245 @@
+                {
+                    clientSession.AbortTransaction(cancellationToken);
+                }
+
+                if (HasErrorLabel(ex, TransientTransactionErrorLabel) && !HasTimedOut(startTime, clock.UtcNow))
+                {
+                    return new CallbackOutcome<TResult>.WithShouldRetryTransaction();
+                }
+
+                throw;
+            }
+        }
+
+        private static async Task<CallbackOutcome<TResult>> ExecuteCallbackAsync<TResult>(IClientSessionHandle clientSession, Func<IClientSessionHandle, CancellationToken, Task<TResult>> callbackAsync, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var result = await callbackAsync(clientSession, cancellationToken).ConfigureAwait(false);
+                return new CallbackOutcome<TResult>.WithResult(result);
+            }
+            catch (Exception ex)
+            {
+                if (IsTransactionInStartingOrInProgressState(clientSession))
+                {
+                    await clientSession.AbortTransactionAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                if (HasErrorLabel(ex, TransientTransactionErrorLabel) && !HasTimedOut(startTime, clock.UtcNow))
+                {
+                    return new CallbackOutcome<TResult>.WithShouldRetryTransaction();
+                }
+
+                throw;
+            }
+        }
+
+        private static bool CommitWithRetries(IClientSessionHandle clientSession, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                try
+                {
+                    clientSession.CommitTransaction(cancellationToken);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    var now = clock.UtcNow;
+                    if (HasErrorLabel(ex, UnknownTransactionCommitResultLabel) && !HasTimedOut(startTime, now))
+                    {
+                        continue;
+                    }
+
+                    if (HasErrorLabel(ex, TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                    {
+                        return false; // the transaction will be retried
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private static async Task<bool> CommitWithRetriesAsync(IClientSessionHandle clientSession, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                try
+                {
+                    await clientSession.CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    var now = clock.UtcNow;
```

Comment:
It would be worth adding a comment:

    var now = clock.UtcNow; // call UtcNow once whether we need to or not to facilitate predictable mocking

Reply:
Done